src/hierarchies/lapnig_hierarchy.cc

```diff
@@ -0,0 +1,200 @@
+      throw std::invalid_argument("mh_log_scale_var parameter must be > 0");
+    }
+  } else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void LapNIGHierarchy::initialize_state() {
+  state.mean = hypers->mean;
+  state.scale = 1. / hypers->scale / (hypers->shape + 1);
+}
+
+void LapNIGHierarchy::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState> &states) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  if (prior->has_fixed_values()) {
+    return;
+  } else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+LapNIG::State LapNIGHierarchy::draw(const LapNIG::Hyperparams &params) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  LapNIG::State out{};
+  out.scale = stan::math::inv_gamma_rng(params.shape, 1. / params.scale, rng);
+  out.mean = stan::math::normal_rng(params.mean, sqrt(params.var), rng);
+  return out;
+}
+
+void LapNIGHierarchy::update_summary_statistics(
+    const Eigen::RowVectorXd &datum, bool add) {
+  if (add) {
+    cluster_data_values.push_back(datum);
+  } else {
+    auto it = std::find(cluster_data_values.begin(), cluster_data_values.end(),
+                        datum);
+    cluster_data_values.erase(it);
+  }
+}
+
+void LapNIGHierarchy::sample_full_cond(bool update_params) {
+  if (this->card == 0) {
+    // No posterior update possible
+    this->sample_prior();
+  } else {
+    // Number of iterations to compute the acceptance rate
+    ++iter_;
+
+    // Random generator
+    auto &rng = bayesmix::Rng::Instance().get();
+
+    // Candidate mean and candidate log_scale
+    Eigen::VectorXd curr_unc_params(2);
+    curr_unc_params << state.mean, std::log(state.scale);
+
+    Eigen::VectorXd prop_unc_params = propose_rwmh(curr_unc_params);
+
+    double log_target_prop = eval_prior_lpdf_unconstrained(prop_unc_params) +
+                             eval_like_lpdf_unconstrained(prop_unc_params);
+
+    double log_target_curr = eval_prior_lpdf_unconstrained(curr_unc_params) +
+                             eval_like_lpdf_unconstrained(curr_unc_params);
+
+    double log_a_rate = log_target_prop - log_target_curr;
+
+    if (std::log(stan::math::uniform_rng(0, 1, rng)) < log_a_rate) {
+      ++accepted_;
+      state.mean = prop_unc_params(0);
+      state.scale = std::exp(prop_unc_params(1));
+    }
+  }
+}
+
+Eigen::VectorXd LapNIGHierarchy::propose_rwmh(
+    const Eigen::VectorXd &curr_vals) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  double candidate_mean =
+      curr_vals(0) + stan::math::normal_rng(0, sqrt(hypers->mh_mean_var), rng);
+  double candidate_log_scale =
+      curr_vals(1) +
+      stan::math::normal_rng(0, sqrt(hypers->mh_log_scale_var), rng);
+  Eigen::VectorXd proposal(2);
+  proposal << candidate_mean, candidate_log_scale;
+  return proposal;
+}
+
+double LapNIGHierarchy::eval_prior_lpdf_unconstrained(
+    Eigen::VectorXd unconstrained_parameters) {
+  return (-0.5 / hypers->var / hypers->var *
```

Comment:
this is rather hard for me to read. Why are you not using the `stan_lpdf` functions?
It would look something like
```
double mu = unconstrained_parameters(0);
double log_scale = unconstrained_parameters(1);
double scale = std::exp(log_scale);

double out = stan::math::normal_lpdf(mu, hypers->mean, hypers->var) + 
  stan::math::inv_gamma_lpdf(scale, hypers ...) + log_det_of_jacobian.
```
also, in the current code you write hypers->var but treat it as a scale, so please be consistent.

Reply:
Okay, we followed your suggestion, thanks. Can you confirm that the usage of hypers->var is consistent now?